docs/extended-tour.md

```diff
@@ -0,0 +1,159 @@
+- [Autogenerated files](extended-tour.md#autogenerated-files)
+- [Overall flow](extended-tour.md#overall-flow)
+
+# Walkthrough example code
+
+This skeleton includes a few examples of how you might design your application, including standalone functions and creating Objects/Classes. Both the synchronous and asynchronous versions of each are included to give an iterative example of what the progression from sync to async looks like. Let's run through reasons why you'd design your code in these ways:
+1. [Standalone function](API.md#hello-1)
+2. [Standalone asynchronous function](../API.md#hello_async)
+3. [Object/Class](../API.md#helloobject)
+4. [Asynchronous Object/Class](../API.md#helloobjectasync)
+
+
+### When would you use a standalone function?
+Standalone functions are simple and useful when your code is doing some kind of basic operation. They avoid the need for a class, since creating a class in this situation would add unnecessary complexity. Creating an asynchronous standalone function can be powerful if your code has a specific, yet expensive operation to do and would benefit from the use of multiple threads at one time to do that expensive work.
+
+### When would you use an object/class?
+Create an object/class when you need to do some kind of data preprocessing before going into the thread pool. It's best to write your code so that the preprocessing happens _once_ as a separate operation, then continues through to the thread pool after the preprocessing is finished and the object is ready. Examples:
+  - [node-mapnik](https://github.com/mapnik/node-mapnik/blob/fe80ce5d79c0e90cfbb5a2b992bf0ae2b8f88198/src/mapnik_map.hpp#L20): we create a Map object once, then use the object multiple times for rendering each vector tile.
+  - [node-addon-examples](https://github.com/nodejs/node-addon-examples/tree/master/6_object_wrap/nan) for another example of what an object-focused Addon looks like.
+
+Objects/Classes are useful when you start doing more complex operations and need to consider performance more heavily, when performance constraints start to matter. Use classes to compute the value of something once, rather than every time you call a function.
+
+One step further is using an asynchronous object or class, which enables you to pass data into the threadpool in the most efficient way possible. This is the main goal of the `HelloObjectAsync` example in this skel, which also uses "move semantics" for even better performance.
+- **Move semantics**: move semantics avoid data being copied (allocating memory) and are most useful when working with big data. `HelloObjectAsync` is using move semantics to limit unnecessary memory allocation, which can be super expensive and generally is one of the main bottle necks of C++ applications. Other scenarios can also cause bottle necks in your code, but memory allocation is the best-case bottle neck to have because you can either control or avoid it altogether.
+
```

Comment:
@GretaCB I've added https://github.com/mapbox/cpp/pull/30 which more details on allocation. It is not absolutely correct to say that allocation is super expensive. Really it is quite fast. The issue is that is is required to create objects on the heap. And most programs need to create a lot of objects on the heap. And once all other obvious bottlenecks have been solved, the remaining bottleneck for many programs I've seen is memory allocation. So, reducing allocations is a final stage optimization to get the fastest code possible. Because it is best to design for efficiency, I think avoiding allocations from the start is a great pattern for writing code even when saving those allocations might not produce measurably faster code in isolated benchmarks. But under high load, when many objects in many threads are being allocated - that is the point that the work of reducing unnecessary allocations will really pay off.

Can you remove the `super expensive` reference and link, as you see fit, to the new glossary items on allocation?